sys/include/net/sock/dtls.h

```diff
@@ -0,0 +1,557 @@
+ * if (!ipv6_addr_from_str((ipv6_addr_t *)remote.addr.ipv6, addr_str)) {
+ *     puts("Error parsing destination address");
+ *     return;
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * After the UDP sock is created, we can proceed with creating the DTLS sock.
+ * Before sending the packet, we must first establish a session with the remote
+ * endpoint using @ref sock_dtls_establish_session(). If the handshake is
+ * successfull and the session is established, we can use its `remote` to send
+ * packets to it with @ref sock_dtls_send().
+ *
+ * If error during any of these operation, the session is closed and the sock
+ * is destroyed using @ref sock_dtls_close_session() and
+ * @ref sock_dtls_destroy().
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * res = sock_dtls_create(&dtls_sock, &udp_sock, DTLS_SOCK_CLIENT_TAG, 0);
+ * if (res < 0) {
+ *     puts("Error creating DTLS sock");
+ *     return;
+ * }
+ *
+ * res = sock_dtls_establish_session(&dtls_sock, &remote, &session);
+ * if (res < 0) {
+ *     printf("Error establishing session: %d\n", (int)res);
+ *     goto end;
+ * }
+ *
+ * res = sock_dtls_send(&dtls_sock, &session, data, datalen);
+ * if (res < 0) {
+ *     printf("Error sending DTLS message: %d\n", (int)res);
+ *     goto end;
+ * }
+ *
+ * res = sock_dtls_recv(&dtls_sock, &session, rcv, sizeof(rcv), SOCK_NO_TIMEOUT);
+ * if (res < 0) {
+ *     printf("Error receiving DTLS message: %d\n", (int)res);
+ *     goto end;
+ * }
+ *
+ * end:
+ * puts("Terminating");
+ * sock_dtls_close_session(&dtls_sock, &session);
+ * sock_dtls_destroy(&dtls_sock);
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * @{
+ *
+ * @file
+ * @brief   DTLS sock definitions
+ *
+ * @author  Aiman Ismail <muhammadaimanbin.ismail@haw-hamburg.de>
+ */
+
+#ifndef NET_SOCK_DTLS_H
+#define NET_SOCK_DTLS_H
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <sys/types.h>
+
+#include "net/sock.h"
+#include "net/sock/udp.h"
+#include "net/credman.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Type for a DTLS sock object
+ *
+ * @note    API implementors: `struct sock_dtls` needs to be defined by
+ *          implementation-specific `sock_dtls_types.h`.
+ */
+typedef struct sock_dtls sock_dtls_t;
+
+/**
+ * @brief Information about the established session with
+ *        the remote endpoint. Used when sending and
+ *        receiving data to the endpoint
+ */
+typedef struct sock_dtls_session sock_dtls_session_t;
+
+/**
+ * @brief Must be called once before any other use
+ */
+void sock_dtls_init(void);
+
+/**
+ * @brief Creates a new DTLS sock object
+ *
+ * @param[out] sock     The resulting DTLS sock object
+ * @param[in] udp_sock  Existing UDP sock to be used
+ * @param[in] tag       Credential tag of the sock. Used to get the right
+ *                      credential from pool.
+ * @param[in] method    Defines the method for the client or server to use.
```

Comment:
What kind of method are you referring to here?

Reply:
In wolfSSL, there is WOLFSSL_METHOD that refers to the whether TLS or DTLS are going to be used, its version (1.0,  1.1, 1.2) and whether the sock is used as client or server. The TLS/DTLS part is irrelevant for DTLS sock, but the version number and client/server part is still relevant.

I kinda ignored it for tinydtls because it only supports DTLS version 1.2 and there is no difference from the start between a server and client. Every tinydtls endpoint is initially a server and a client is an endpoint that calls `dtls_connect()` later on.

So, I'm not sure what to name the parameter here to describe that "this parameter determines which DTLS version to use and functionality of this sock (client/server)".

Reply:
Maybe rename it version? In any case there need to be some set of defines or enum, that can be given to the API here. Having the value (or at least its variable name) be implementation-dependent would break the abstraction of the API and require the user to always fiddle around in the code if they switch libraries (which is the whole point we try to prevent here).

Reply:
Okay, will do.

Reply:
> In any case there need to be some set of defines or enum, that can be given to the API here.

I just saw, that https://github.com/RIOT-OS/RIOT/issues/10897 mentioned those, so apparently they were in our original plans ;-)

Reply:
I'm referring to

> ```C
> /**
>  * Methods for sock_dtls_create.
>  * Defines the DTLS protocol version to use.
>  */
> #define DTLSv1_SERVER_METHOD
> #define DTLSv1_CLIENT_METHOD
> #define DTLSv12_SERVER_METHOD
> #define DTLSv12_CLIENT_METHOD
> 

```

Reply:
Alternatively, we can separate the DTLS version from role of the sock. For setting the DTLS version, I'm thinking of adding a new function  to configure any other configuration options offered by the underlying DTLS implementation. More on that in this [comment](https://github.com/RIOT-OS/RIOT/pull/11909#issuecomment-519114289)

Aside from the naming, we actually know when creating the DTLS sock, which role (client/server) the sock will take so any function calls required by the DTLS implementation to start a DTLS server can be done directly in `sock_dtls_create()`. Therefore `sock_dtls_init_server()` is not needed anymore and can be removed.

```C
/**
 * @brief DTLS role for the sock
 */
typedef enum {
    SOCK_DTLS_IS_SERVER,      /* DTLS server role */
    SOCK_DTLS_IS_CLIENT,      /* DTLS client role */
} sock_dtls_role_t;

[...]

/**
 * @brief Creates a new DTLS sock object
 *
 * @see net_credman.
 *
 * @param[out] sock     The resulting DTLS sock object
 * @param[in] udp_sock  Existing UDP sock to be used underneath
 * @param[in] tag       Credential tag of the sock. The sock will only use
 *                      credentials with the same tag given here.
 * @param[in] role      DTLS role of the sock.
 *
 * @return  0 on success.
 * @return  -1 on error
 */
int sock_dtls_create(sock_dtls_t *sock, sock_udp_t *udp_sock,
                     credman_tag_t tag, sock_dtls_role_t role);
```

Reply:
> Alternatively, we can separate the DTLS version from role of the sock. For setting the DTLS version, I'm thinking of adding a new function to configure any other configuration options offered by the underlying DTLS implementation. More on that in this [comment](https://github.com/RIOT-OS/RIOT/pull/11909#issuecomment-519114289)

Since the versions are incompatible to each other, I think it is important to cover them in this API, beyond your configuration proposal

Reply:
Added commit to address this. See [comment](https://github.com/RIOT-OS/RIOT/pull/11909#issuecomment-519494942).

Reply:
Added commit to address this. See [comment](https://github.com/RIOT-OS/RIOT/pull/11909#issuecomment-519494942).